core/src/main/java/com/linecorp/armeria/client/AbstractHttpRequestHandler.java

```diff
@@ -230,9 +244,37 @@ final void writeHeaders(RequestHeaders headers) {
         // Attach a listener first to make the listener early handle a cause raised while writing headers
         // before any other callbacks like `onStreamClosed()` are invoked.
         promise.addListener(this);
-        encoder.writeHeaders(id, streamId(), merged, headersOnly, promise);
+        encoder.writeHeaders(id, streamId(), headers, headersOnly, promise);
     }
 
+    static boolean needs100Continue(RequestHeaders headers) {
+        return headers.contains(HttpHeaderNames.EXPECT, HttpHeaderValues.CONTINUE.toString());
+    }
+
+    void handle100Continue(ResponseHeaders responseHeaders) {
+        if (state != State.NEEDS_100_CONTINUE) {
+            return;
+        }
+
+        if (responseHeaders.status() == HttpStatus.CONTINUE) {
+            state = State.NEEDS_DATA_OR_TRAILERS;
+            resume();
+            // TODO(minwoox): reset the timeout
+        } else {
+            // We do not retry the request when HttpStatus.EXPECTATION_FAILED is received
```

Comment:
Q) If the request was an improper one, wouldn't it be better to raise an exception such as `ExpectionFailedException`? The exception could be used when implementing retrying logic.

Reply:
That's a good suggestion. 👍

Reply:
On second thought, I'm not sure if it's a good idea. We don't raise an `InternalServerException` when we receive 500. Can't we just let the user handle the response depending on the response status?

Reply:
Makes sense. Let's keep it as is.